crates/services/parallel-executor/src/validator/dependency_graph.rs

```diff
@@ -0,0 +1,522 @@
+use std::collections::{
+    HashMap,
+    HashSet,
+    VecDeque,
+};
+
+use fuel_core_types::{
+    blockchain::transaction::TransactionExt,
+    fuel_tx::{
+        ContractId,
+        Input,
+        Transaction,
+    },
+};
+
+/// Dependency graph for contract-based transactions
+#[derive(Debug, Default)]
+pub struct DependencyGraph {
+    /// Contract ID -> Set of transaction indices that use this contract
+    contract_to_transactions: HashMap<ContractId, HashSet<usize>>,
+    /// Transaction index -> Set of contract IDs it depends on
+    transaction_to_contracts: HashMap<usize, HashSet<ContractId>>,
+    /// Transaction index -> Transaction
+    transactions: HashMap<usize, Transaction>,
+    /// Tracks remaining dependencies for each transaction
+    remaining_dependencies: HashMap<usize, usize>,
+    /// Cache of currently independent transactions
+    independent_transactions: HashSet<usize>,
+    /// Queue for newly independent transactions
+    newly_independent_queue: VecDeque<usize>,
```

Comment:
I don't understand why we need this in addition of `independent_transactions` is it not possible to be only vecdeque ?

Reply:
for O(1) lookup and popping :)